nanoFramework.Device.Bluetooth/IO/DataReader.cs

```diff
@@ -0,0 +1,296 @@
+﻿//
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Text;
+
+namespace nanoFramework.Device.Bluetooth
+{
+    /// <summary>
+    /// Reads data from an input buffer.
+    /// </summary>
+    public sealed class DataReader 
+    {
+        private Buffer _buffer;
+        private int _currentReadPosition;
+
+        const int defaultBufferSize = 32;
+
+        private DataReader(Buffer buffer)
+        {
+            _buffer = buffer;
+            _currentReadPosition = 0;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the data reader with data from the specified buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer.</param>
+        /// <returns>The data reader.</returns>
+        public static DataReader FromBuffer(Buffer buffer)
+        {
+            return new DataReader(buffer);
+        }
+
+        /// <summary>
+        /// Gets the size of the buffer that has not been read.
+        /// </summary>
+        /// <value>
+        /// The size of the buffer that has not been read, in bytes.
+        /// </value>
+        public uint UnconsumedBufferLength { get { return (_buffer.Length - (uint)_currentReadPosition); } }
+
+
+        /// <summary>
+        /// Reads a Boolean value from the input buffer.
+        /// </summary>
+        /// <returns>The value.</returns>
+        public bool ReadBoolean()
+        {
+            var value = _buffer.Data[IncreaseReadPosition(1)] > 0;
```

Comment:
general question for all those function, is if worse checking that you have the needed bytes to read before the conversion or reading the buffer?

Reply:
That's done in IncreaseReadPosition(?) and raises a ArgumentOutOfRangeException 
This is mainly existing code from Windows.Storage.Streams